drv/lpc55-update-server/src/images.rs

```diff
@@ -26,79 +47,617 @@ extern "C" {
+        let vectors = match self.accessor {
+            Accessor::Flash { .. } => {
+                let buffer =
+                    &mut [0u8; core::mem::size_of::<ImageVectorsLpc55>()];
+                self.read_bytes(0u32, buffer.as_bytes_mut())?;
+                ImageVectorsLpc55::read_from_prefix(&buffer[..])
+                    .ok_or(UpdateError::OutOfBounds)
+            }
+            Accessor::Ram { buffer, .. } | Accessor::Hybrid { buffer, .. } => {
+                ImageVectorsLpc55::read_from_prefix(buffer)
+                    .ok_or(UpdateError::OutOfBounds)
+            }
+        }?;
+        let len = vectors.image_length().ok_or(UpdateError::BadLength)?;
+        round_up_to_flash_page(len).ok_or(UpdateError::BadLength)
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub struct Epoch {
+    value: u32,
+}
+
+/// Convert from the ImageHeader.epoch format
+impl From<u32> for Epoch {
+    fn from(number: u32) -> Self {
+        Epoch { value: number }
+    }
+}
+
+/// Convert from the caboose EPOC value format.
+//
+// Invalid EPOC values converted to Epoch{value:0} include:
+//   - empty slice
+//   - any non-ASCII-digits in slice
+//   - any non-UTF8 in slice
+//   - leading '+' normally allowed by parse()
+//   - values greater than u32::MAX
+//
+//   Hand coding reduces size by about 950 bytes.
+impl From<&[u8]> for Epoch {
+    fn from(chars: &[u8]) -> Self {
+        let epoch =
+            if chars.first().map(|c| c.is_ascii_digit()).unwrap_or(false) {
+                if let Ok(chars) = core::str::from_utf8(chars) {
+                    chars.parse::<u32>().unwrap_or(0)
+                } else {
+                    0
+                }
+            } else {
+                0
+            };
+        Epoch::from(epoch)
+    }
+}
+
+impl Epoch {
+    pub fn can_update_to(&self, next: Epoch) -> bool {
+        self.value <= next.value
+    }
+}
+
+/// Check a next image against an active image to determine
+/// if rollback policy allows the next image.
+/// If ImageHeader and Caboose are absent or Caboose does
+/// not have an `EPOC` tag, then the Epoch is defaulted to zero.
+/// This test is also used when the header block first arrives
+/// so that images can be rejected early, i.e. before any flash has
+/// been altered.
+pub fn check_rollback_policy(
+    active_image: ImageAccess<'_>,
+    next_image: ImageAccess<'_>,
+    complete: bool,
+) -> Result<(), UpdateError> {
+    let next_epoch = get_image_epoch(&next_image)?;
+    let active_epoch = get_image_epoch(&active_image)?;
+    match (active_epoch, next_epoch) {
+        // No active_epoch is treated as zero; update can proceed.
+        (None, _) => Ok(()),
+        (Some(active_epoch), None) => {
+            // If next_image is partial and HEADER_BLOCK has no ImageHeader,
+            // then there is no early rejection, proceed.
+            if !complete || active_epoch.can_update_to(Epoch::from(0u32)) {
```

Comment:
Same vague concerns as before: being able to update to an invalid / missing epoch (but only from epoch 0) seems odd to me.